aptos-move/framework/aptos-framework/sources/configs/Stake.move

```diff
@@ -0,0 +1,247 @@
+        // sum of active and cooldown stakes, updated on epoch boundary.
+        current_stake: u64,
+        // active stake
+        active: vector<Delegation>,
+        // inactive stake, can be withdrawn
+        inactive: vector<Delegation>,
+        // pending activation for next epoch
+        warmup: vector<Delegation>,
+        // pending deactivation for next epoch
+        cooldown: vector<Delegation>,
+    }
+
+    /// Consensus information per validator, stored in validator address.
+    struct ValidatorInfo has key {
+        stake_info: StakeInfo,
+        consensus_pubkey: vector<u8>,
+        network_address: vector<u8>,
+    }
+
+    /// Full ValidatorSet, stored in @CoreResource.
+    struct ValidatorSet has key {
+        consensus_scheme: u8,
+        // minimum stakes requires to join validator set
+        minimum_stake: u64,
+        // maximum stakes allowed to join validator set
+        maximum_stake: u64,
+        // should we store the ValidatorInfo in place?
+        validators: vector<address>,
+        last_update_time_secs: u64,
+    }
+
+    /// Any user can delegate a stake.
+    fun delegate_stake(account: &signer, to: address, coin: Coin, locked_until_secs: u64) acquires ValidatorInfo, ValidatorSet {
+        let current_time = Timestamp::now_seconds();
+        assert!(current_time + MINIMUM_LOCK_PERIOD < locked_until_secs, 0);
+        let stake_info = &mut borrow_global_mut<ValidatorInfo>(to).stake_info;
+        let delegation = Delegation {
+            coin,
+            locked_until_secs,
+            from: Signer::address_of(account),
+        };
+        // add to warmup if it's a current validator otherwise add to active directly
+        if (is_validator(to)) {
+            Vector::push_back(&mut stake_info.warmup, delegation);
+        } else {
+            stake_info.current_stake = stake_info.current_stake + TestCoin::value(&delegation.coin);
+            Vector::push_back(&mut stake_info.active, delegation);
+        }
+    }
+
+    /// Withdraw from active delegation, it's moved to cooldown if locked_until_secs < current_time or
+    /// directly deposit if it's not from an active validator.
+    fun withdraw_active(account: &signer, from: address) acquires ValidatorInfo, ValidatorSet {
+        let addr = Signer::address_of(account);
+        let current_time = Timestamp::now_seconds();
+        let stake_info = &mut borrow_global_mut<ValidatorInfo>(from).stake_info;
+        let d = withdraw_internal(&mut stake_info.inactive, addr);
+        let is_validator = is_validator(from);
+        if (!is_validator) {
+            // directly deposit if it's not active validator
+            let Delegation {coin, from: _, locked_until_secs: _} = d;
+            TestCoin::deposit(addr, coin);
+        } else if (d.locked_until_secs < current_time) {
+            // move to cooldown if it can be unlocked
+            Vector::push_back(&mut stake_info.cooldown, d);
+        } else {
```

Comment:
it feels like someone shouldn't be able to extract their delegation unless the time to remove has lapsed. This leads to some quirky and unpredictable behavior for humans